graylog2-web-interface/src/views/components/contexts/WidgetFocusProvider.tsx

```diff
@@ -15,67 +15,123 @@
  * <http://www.mongodb.com/licensing/server-side-public-license>.
  */
 import * as React from 'react';
-import { useState, useEffect, useCallback } from 'react';
+import { isEqual } from 'lodash';
+import { useState, useEffect, useCallback, useMemo } from 'react';
 import { useLocation, useHistory } from 'react-router-dom';
 import PropTypes from 'prop-types';
 import URI from 'urijs';
 
 import { useStore } from 'stores/connect';
 import useQuery from 'routing/useQuery';
 import { WidgetStore } from 'views/stores/WidgetStore';
-import WidgetFocusContext from 'views/components/contexts/WidgetFocusContext';
+import WidgetFocusContext, { FocusRequest } from 'views/components/contexts/WidgetFocusContext';
 
-const _syncWithQuery = (query: string, focusedWidget: string) => {
-  const baseUri = new URI(query)
-    .removeSearch('focused');
+type QueryParamsUpdate = FocusRequest & { id: FocusRequest['id'] | undefined };
```

Comment:
I think we can define the different possible states (focused widget, no focused widget, editing widget) here to make invalid states impossible.

Reply:
I've added explicit types for the context state and the query params update. We need slightly more code for the query params update, but it also improved the readability.